pkg/transports/connecting/dtls/nat.go

```diff
@@ -75,55 +79,57 @@ func openUDPLimitTTL(ctx context.Context, laddr, addr string, dialer dialFunc) e
 	return nil
 }
 
-var (
-	privPortSingle int
-	pubPortSingle  int
-)
+func publicAddr(ctx context.Context, network string, stunServer string, dialer dialFunc) (privateAddr *net.UDPAddr, publicAddr *net.UDPAddr, err error) {
 
-func publicAddr(stunServer string, dialer func(ctx context.Context, network, laddr, raddr string) (net.Conn, error)) (privatePort int, publicPort int, err error) {
-
-	if privPortSingle != 0 && pubPortSingle != 0 {
-		return privPortSingle, pubPortSingle, nil
-	}
-
-	udpConn, err := dialer(context.Background(), "udp", "", stunServer)
+	udpConn, err := dialer(ctx, network, "", stunServer)
 	if err != nil {
-		return 0, 0, fmt.Errorf("error connecting to STUN server: %v", err)
+		return nil, nil, fmt.Errorf("error connecting to STUN server: %v", err)
 	}
 	defer udpConn.Close()
 
 	localAddr, err := net.ResolveUDPAddr(udpConn.LocalAddr().Network(), udpConn.LocalAddr().String())
 	if err != nil {
-		return 0, 0, fmt.Errorf("error resolving local address: %v", err)
+		return nil, nil, fmt.Errorf("error resolving local address: %v", err)
 	}
 
 	client, err := stun.NewClient(udpConn)
 	if err != nil {
-		return 0, 0, fmt.Errorf("error creating STUN client: %v", err)
+		return nil, nil, fmt.Errorf("error creating STUN client: %v", err)
 	}
 
 	message := stun.MustBuild(stun.TransactionID, stun.BindingRequest)
 
 	var xorAddr stun.XORMappedAddress
 
-	err = client.Do(message, func(res stun.Event) {
+	doneCh := make(chan error)
```

Comment:
Add a buffer of size 1 to `doneCh` to ensure the following sends won't block forever, leaving dangling goroutines. This can happen when the `<-ctx.Done()` select case proceeds first, leaving no receiver for `doneCh`.

Reply:
Good catch! Added, thanks for looking into this!